clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java

```diff
@@ -253,6 +284,58 @@ public void forceClose() {
         initiateClose();
     }
 
+    private void sendInitPidRequest(Node node) {
+        String nodeId = node.idString();
+        InitPidRequest.Builder builder = new InitPidRequest.Builder(null);
+        RequestCompletionHandler callback = new RequestCompletionHandler() {
+            public void onComplete(ClientResponse response) {
+                handleInitPidResponse(response);
+            }
+        };
+        ClientRequest request = client.newClientRequest(nodeId, builder, time.milliseconds(), true, callback);
+        client.send(request, time.milliseconds());
+    }
+
+    private Node getReadyNode(long remainingTimeMs) {
+        Node node = client.leastLoadedNode(time.milliseconds());
+        if (node == null) {
+            return null;
+        }
+        client.ready(node, time.milliseconds());
+        if (client.connectionFailed(node))
+            return null;
+        client.poll(remainingTimeMs, time.milliseconds());
+        if (client.isReady(node, time.milliseconds()))
+            return node;
+        return null;
+    }
+
+    private void maybeWaitForPid() {
+        while (!transactionState.hasPid()) {
+            try {
+                Node node = getReadyNode(requestTimeout);
+                if (node == null) {
+                    log.warn("Could not find an available node while trying to get a producer id.");
+                    throw new NoAvailableBrokersException();
```

Comment:
Maybe we should simply backoff (i.e. sleep for the retry backoff duration) and try again since this is the background thread and we cannot propagate exceptions. I think we usually expect this to be a transient state anyway (we assume a broker will eventually come online).